File: problems/685_redundant_connection_ii/solution.py
class Solution:
    def findRedundantDirectedConnection(self, edges: list[list[int]]) -> list[int]:
        n = len(edges)
        parent = [0] * (n + 1)
        cand1 = cand2 = None
        for u, v in edges:
            if parent[v] != 0:
                cand1 = [parent[v], v]
                cand2 = [u, v]
                break
            parent[v] = u
        root = list(range(n + 1))
        def find(x):
            while root[x] != x:
                root[x] = root[root[x]]
                x = root[x]
            return x
        for u, v in edges:
            if cand2 and u == cand2[0] and v == cand2[1]:
                continue
            pu, pv = find(u), find(v)
            if pu == pv:
                return cand1 if cand1 else [u, v]
            root[pv] = pu
        return cand2

File: problems/685_redundant_connection_ii/Solution.java
class Solution {
    public int[] findRedundantDirectedConnection(int[][] edges) {
        int n = edges.length;
        int[] parent = new int[n + 1];
        int[] cand1 = null, cand2 = null;
        for (int[] e : edges) {
            if (parent[e[1]] != 0) {
                cand1 = new int[]{parent[e[1]], e[1]};
                cand2 = new int[]{e[0], e[1]};
                break;
            }
            parent[e[1]] = e[0];
        }
        int[] root = new int[n + 1];
        for (int i = 0; i <= n; i++) root[i] = i;
        for (int[] e : edges) {
            if (cand2 != null && e[0] == cand2[0] && e[1] == cand2[1]) continue;
            int pu = find(root, e[0]), pv = find(root, e[1]);
            if (pu == pv) return cand1 != null ? cand1 : e;
            root[pv] = pu;
        }
        return cand2;
    }
    private int find(int[] root, int x) {
        while (root[x] != x) { root[x] = root[root[x]]; x = root[x]; }
        return x;
    }
}
